### PR TITLE
Add DNS pod resource monitoring option

### PR DESCRIPTION
**Pull request by chrisohaver**
**What this PR does / why we need it**:

Adds option for tracking of DNS pods + master pods stats.

**Special notes for your reviewer**:

Relates to https://github.com/kubernetes/kubernetes/issues/68613, in which would be ideal to have the option to track DNS pod resource and master pods.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by chrisohaver** [comment]
@wojtek-t, PTAL.  I have not tested this yet (I have not figured that part out yet).  But I'm hoping you can review this to see if this is on the right track.

**Comment by nikopen** [comment]
@neolit123 this needs

/milestone v1.12
/priority critical-urgent

to be merged due to code freeze

**Comment by fturib** [comment]
I launch test on PR that exercice the option --gather-resource-usage=master
Here on the 500 nodes perf use case.
I guess we should then verify that we have the monitoring collected in the artifacts of the tests, I think in the files "ResourceUsageSummary xxxx". (we would need verify we have info of master and nodes that host the DNS).

/test pull-kubernetes-e2e-gce-big-performance

**Review by fturib — COMMENTED** [review]
on line #276 : if you expect to monirtor several noeds even though that is a MasterOnly option, then I guess we should not break, but just continue to match another node of the dnsNodes list.

**Comment by fturib** [comment]
FYI : the non required test I requested "pull-jubernetes-e2e-gce-big-performance" is failing. I do not know why. But the reason I added is to check if the nodes where DNS server is running are monitored.

And if I look in the details of that test, It looks like no (see artifacts / ResourceUsageSummary ).

I provided a proposition by comment on the code change.

**Comment by fturib** [comment]
@BenTheElder : I think it is not really a timeout, but rather that the constrainst on CPU is not in the expected range :` l7-lb-controller is using 0.305286849/0.275 CPU`
And that may not be a big deal as we are running on pre-submit here. I mean the cluster is maybe not tune the same way as for the peer test not presubmit.

**Comment by BenTheElder** [comment]
The job did time out, what I linked to is a `kubetest` timeout. It looks like it _also_ didn't pass the CPU constraint.

**Comment by BenTheElder** [comment]
Also yeah, that job hasn't run in presubmit in long (enough that we have no previous entries stored) and it's only the 9th run ever so I'm not sure how worried I'd be anyhow.

**Comment by wojtek-t** [comment]
> @BenTheElder : I think it is not really a timeout, but rather that the constrainst on CPU is not in the expected range :l7-lb-controller is using 0.305286849/0.275 CPUAnd that may not be a big deal as we are running on pre-submit here. I mean the cluster is maybe not tune the same way as for the peer test not presubmit.

I wouldn't worry that much about this problem. Those thresholds were adjusted for some scales, but not for 500-node ones. So I think it's fine.


About the monitoring DNS resources - it doesn't seem to work.

File: test/e2e/framework/resource_usage_gatherer.go [hunk 1]
```diff
@@ -237,13 +237,20 @@ func NewResourceUsageGatherer(c clientset.Interface, options ResourceGathererOpt
...
+				continue
+			}
 			for _, container := range pod.Status.InitContainerStatuses {
 				g.containerIDs = append(g.containerIDs, container.Name)
 			}
 			for _, container := range pod.Status.ContainerStatuses {
 				g.containerIDs = append(g.containerIDs, container.Name)
 			}
+			if options.MasterOnly {
+				dnsNodes[pod.Spec.NodeName] = true
```
Comment (wojtek-t) [inline test/e2e/framework/resource_usage_gatherer.go]: I'm not a fan of this.
DNS pods are not running on master VM. So hacking them into "MasterOnly" seems wrong to me.

Instead the MasterOnly option should be changed to some enum, with values:
MasterOnly
MasterAndDNS
Everything
[better names appreciated]

**Comment by chrisohaver** [comment]
> About the monitoring DNS resources - it doesn't seem to work.

@wojtek-t, Any idea on why that may be?  Do the changes look right (other than naming nits)?  Or am I completely off base?

**Comment by chrisohaver** [comment]
> on line #276 : if you expect to monirtor several noeds even though that is a MasterOnly option, then I guess we should not break, but just continue to match another node of the dnsNodes list.

Ah - yes i forgot to remove that, thanks!

File: test/e2e/framework/resource_usage_gatherer.go (hunk 1 above)
Comment (chrisohaver) [reply test/e2e/framework/resource_usage_gatherer.go]: OK renamed to enum ...  

AllNodes
MasterNodes
MasterAndDNSNodes

**Comment by chrisohaver** [comment]
This PR, and #9445 are ready for review.  Both are required to track DNS resource usage in the scale tests.  Thanks!

**Comment by wojtek-t** [comment]
/approve

@shyamjvs - any comments?

File: test/e2e/framework/framework.go [hunk 2]
```diff
@@ -247,9 +247,18 @@ func (f *Framework) BeforeEach() {
 
 	if TestContext.GatherKubeSystemResourceUsageData != "false" && TestContext.GatherKubeSystemResourceUsageData != "none" {
 		var err error
+		var nodeMode NodesSet
+		switch TestContext.GatherKubeSystemResourceUsageData {
+		case "master":
+			nodeMode = MasterAndDNSNodes
```
Comment (shyamjvs) [inline test/e2e/framework/framework.go]: Shouldn't this be just master?

Comment (chrisohaver) [reply test/e2e/framework/framework.go]: That’s just a tenporary change to test the collection inside this PR only. If it works, Next we will add coredns as default in kubeup in this PR (and als bump memory limit), and rerun the scale test again to observe coreDNS memory usage.

**Comment by chrisohaver** [comment]
> That’s just a temporary change to test the collection inside this PR only. If it works, Next we will add coredns as default in `kubeup` in this PR (and also bump coredns memory limit), and rerun the scale test again to observe coreDNS memory usage. 

... Also, when we are done with this testing, it's up to you if you want to continue tracking DNS metrics.  If so, I'll of course change that line back to `MasterNodes` and we can merge this (along with kubernetes/test-infra#9445). If not, we can discard the PRs.

**Comment by chrisohaver** [comment]
Metrics collection for DNS pods appears to have worked.  Latest commit is a test commit, to allow testing with CoreDNS with memory bump at 500 nodes to compare to kube-dns.
CoreDNS test in progress now ... pull-kubernetes-e2e-gce-big-performance.

**Comment by chrisohaver** [comment]
Per `pull-kubernetes-e2e-gce-big-performance` results, kube-dns appears to use ~40% more memory than CoreDNS in a 500 node load test.

CoreDNS @ 500 nodes: ~35M per replica
kube-dns @ 500 nodes: ~49M per replica

Now Running tests at 2000 nodes for CoreDNS (this test takes many hours).

**Comment by wojtek-t** [comment]
So given the discussion over email:

- I'm fine with merging this PR temporarily to get confirmation on numbers
- the last two commits (in particular enabling coreDNS by default) will be reverted after the test run (so tomorrow)

We can't afford bumping DNS requests (even by 50m) - this has a big potential to break users as it will preempt their workloads if their clusters are fully packed.

@thockin @tpepper @kubernetes/sig-network-bugs ^^

/lgtm
/milestone v1.12

**Comment by wojtek-t** [comment]
/hold cancel

> have one comment here. Instead of changing the default here (which would change it everywhere), can you just set the env var in our 5k node job config?

We will - but we want to have those results ASAP to not slow down the release. And we need to have this merged O(now).

**Comment by wojtek-t** [comment]
/hold cancel

Actually, I chatted with @thockin and since we know we are not going to put CoreDNS back (because it requires more resources), then it's safer to first cut the branch and only then merge this.

**Comment by shyamjvs** [comment]
Ok, that sounds quite reasonable.

**Comment by nikopen** [comment]
@wojtek-t is there any particular reasoning behind the resource limit on the manifest? I imagine the decision to put the limit on kube-dns was made quite a while ago, and CoreDNS copy/pasted the same numbers.

As a cluster critical component an OOM behaviour is undesirable, if there is any kind of abnormality it should be preferable to allow bursts, thus not having a specified limit at all.

So a 50mb usage difference at the scale of nearly 5k nodes is extremely meagre, as I cannot imagine such large clusters being at 99.9% usage of allocatable resources (quite rare, if not just impossible).

Let's also bear in mind that currently both the metrics-server and the kubelet/eviction-manager only see the 'working set' memory which includes large portions of cache, and in most scenarios RSS memory is much less than what k8s sees.

Might be a bit late for CoreDNS to make it for this release now but it would be interesting to see the (working set) numbers.

**Comment by nikopen** [comment]
In my post above I mentioned removing the limit altogether - 
i.e. what's the reasoning to having a memory limit at all?

**Comment by fturib** [comment]
On the other hand, as it is merged, maybe we can just wait the result of the e2e scale-performance before reverting: we will have the expected result then ...

**Comment by chrisohaver** [comment]
@nikopen, IIUC, removing the limit would default it to 512mi.  It would still have a limit.

@wojtek-t, I am trying to find a description of how the defined memory limit (not actual memory used) factors into scheduling.  Or does it affect clusters in some other way than scheduling? Can you point me to docs/resource?

**Pull request merged**